### ISSUE TITLE
OS X Lion and Python 3.11.3 : make failed

### ISSUE THREAD

**Issue by Laurent-777** [python/cpython#104127]
Hi : 
Thank you for studying the issue I had with Python 3.11.3 tied to be built under Mac OS X Lion . Hereafter the details provided to configure : 
Mac OS X : Lion (10.7.5)
Hardware : Intel Core 2 Duo (X86_64)
Compiler(s) used : GN GCC/G++ 7.5 (or 4.9)
Last version of Python 3 sucessfully built and installed : 3.10.7 (NB Later versions failed in the building process).
Expected result : everything OK (NB Before I buy a new Mac computer)

Obtained result : 
1/ configure is OK (see the log file)
2/ configure displayed adequately the output of its process (see the log file)
3/ make failed (see the log file), with the following message (shortly) : 

```plain
Objects/longobject.c: In function ‘bit_length_digit’:
Objects/longobject.c:779:5: error: implicit declaration of function ‘static_assert’ [-Werror=implicit-function-declaration]
     static_assert(PyLong_SHIFT <= sizeof(unsigned long) * 8,
     ^~~~~~~~~~~~~
/Volumes/External_7/_Ze-Bins/_GNU-Compilers/_7.5.x/bin/gcc -c  -DNDEBUG -g -fwrapv -O3 -Wall -O3   -std=c11 -Werror=implicit-function-declaration -fvisibility=hidden  -I./Include/internal  -I. -I./Include -I/Volumes/External_7/_Ze-Libraries/_GNU-ncurses/include				-I/Volumes/External_7/_Ze-Libraries/_GNU-readline/include				-I/Volumes/External_7/_Ze-Libraries/_GNU-gettext/include				-I/Volumes/External_7/_Ze-Libraries/_FFI-libffi/lib/libffi-3.2.1/include			-I/Volumes/External_7/_Ze-Libraries/_EXPAT/include			-I/Volumes/External_7/_Ze-Bins/_XZ/include			-I/Volumes/External_7/_Ze-Bins/_DB-GDBM/include	   -DPy_BUILD_CORE -o Objects/dictobject.o Objects/dictobject.c
/Volumes/External_7/_Ze-Bins/_GNU-Compilers/_7.5.x/bin/gcc -c  -DNDEBUG -g -fwrapv -O3 -Wall -O3   -std=c11 -Werror=implicit-function-declaration -fvisibility=hidden  -I./Include/internal  -I. -I./Include -I/Volumes/External_7/_Ze-Libraries/_GNU-ncurses/include				-I/Volumes/External_7/_Ze-Libraries/_GNU-readline/include				-I/Volumes/External_7/_Ze-Libraries/_GNU-gettext/include				-I/Volumes/External_7/_Ze-Libraries/_FFI-libffi/lib/libffi-3.2.1/include			-I/Volumes/External_7/_Ze-Libraries/_EXPAT/include			-I/Volumes/External_7/_Ze-Bins/_XZ/include			-I/Volumes/External_7/_Ze-Bins/_DB-GDBM/include	   -DPy_BUILD_CORE -o Objects/odictobject.o Objects/odictobject.c
cc1: some warnings being treated as errors
my-make: *** [Makefile:2459: Objects/longobject.o] Error 1
my-make: *** Waiting for unfinished jobs....
```

Thank you !

Encl. 3 log/text files
[config.log](https://github.com/python/cpython/files/11381928/config.log)
[my-configure.log](https://github.com/python/cpython/files/11381930/my-configure.log)
[my-make.log](https://github.com/python/cpython/files/11381932/my-make.log)

**Comment by sobolevn** [comment]
I don't think that there's a reason for us to support Lion. It is a 12-year old unsupported-by-Apple OS.

**Comment by ned-deily** [comment]
This problem is a duplicate of #99069; fixes were merged for release in the upcoming 3.12.0b1 and 3.11.4 releases.

As noted above, the cPython project does no regular testing of or provides guarantees of support for such old versions of operating systems like macOS 10.7. We may accept PRs for them depending on the impact to the current code base. A suggestion, if you are not already aware of it: the [MacPorts project](https://www.macports.org/install.php) provides support for many open source projects including the latest versions of Python and its dependencies on many older versions of macOS. In fact, the fix for this problem was supplied by MacPorts. They even provide binary installers and many pre-compiled binary packages for 10.7 Lion, including Python 3.11.3 and more recent compilers. I was able to install MacPorts and its Python 3.11.3 on a 10.7 system in about 5 minutes.

**Comment by Laurent-777** [comment]
Good morning : 

Currently investigating the "bug" I have. It seems that ancompiler option make the process failing (TO BE GIVEN LATER -Werror=implicit- etc). I found this opiton in the configure file. I changed it in -Wimplicit etc. Fixed, but a new error appears (pragma ...).

Neverthless, everything is OK with PYTHON3 3.10.9, 3.10.10, 3.10.11 compiled with GNU GCC 4.9.x, 7.5.x, and the all new 13.1.0 (all the sources dowloaded from ftp gnu). 

AN IMPORTANT HINT !

In fact, the C++ compiler provided by the Lion-based XCode, namely, Apple clang, doesn't work with the last versions of 3.10.x. I found it, because without setting CXX for configure, configure looks for a C++ inside /usr/bin... that is Apple clang. Ho wto fix it ? In setting CXX as the C++ 4.9.x, 7.5.x, or 13.1.0 as an example. 
I looked for the log file issued by ./configure --help. I didn't notice CXX (just a line mention a C++ compielr, but not the usual CXX). Maybe one need to modify it.

Thank you very much, and have a good day !
